File: 2020/day14/part1.js
const path = require('path')
const readFile = require('../../utils/read-file')

const maskPrefix = 'mask = '
const allocationPrefix = 'mem'

const decode = async () => {
  const data = await readFile(path.resolve(__dirname, 'data.txt'))
  const values = data.toString().trim().split('\n')
  const memory = saveToMemory(values)
  const result = memory.reduce((x, y) => x + y.value, 0)
  console.log(result)
}

const saveToMemory = (values) => {
  const memory = []
  let mask
  for (let i = 0; i < values.length; i++) {
    if (values[i].startsWith(maskPrefix)) {
      mask = parseMask(values[i])
    } else if (values[i].startsWith(allocationPrefix)) {
      const allocation = parseAllocation(values[i])
      const valueBinary = convertValueToBinaryArray(allocation[1])
      mask.forEach(change => {
        valueBinary[change.position] = change.value
      })
      const updatedValue = convertBinaryArrayToValue(valueBinary)
      const index = memory.findIndex(x => x.location === allocation[0])
      if (index !== -1) {
        memory[index].value = updatedValue
      } else {
        memory.push({ location: allocation[0], value: updatedValue })
      }
    }
  }
  return memory
}

function parseMask (value) {
  return value.substring(maskPrefix.length).split('').map((x, i) => { return { value: x, position: i } }).filter(x => x.value !== 'X')
}

function parseAllocation (value) {
  return value.replace('mem[', '').replace('] =', '').split(' ')
}

function convertValueToBinaryArray (value) {
  return Number(value).toString(2).padStart(36, '0').split('')
}

function convertBinaryArrayToValue (array) {
  return parseInt(array.join(''), 2)
}

(async function () {
  await decode()
}())

File: 2021/day4/part1.py
import os
f = open(os.path.join(os.path.dirname(__file__), 'data.txt'))
data = f.read().split('\n\n')

balls = data[0].replace('\n', '').split(',')

tickets = []

for i in range(len(data)):
    if i > 0:
        ticket = data[i].splitlines()
        for i, row in enumerate(ticket):
            ticket[i] = filter(lambda x: x != '', row.split(' '))
        tickets.append(ticket)


def draw_balls():
    for ball in balls:
        for ticket in tickets:
            win = False
            for row in ticket:
                if ball in row:
                    index = row.index(ball)
                    row[index] = 'X'

                if len(row) == len(filter(lambda x: (x == 'X'), row)):
                    win = True

            for y in range(len(ticket[0])):
                column_checked = 0

                for z in range(len(ticket)):
                    if ticket[z][y] == 'X':
                        column_checked += 1

                if column_checked == len(ticket):
                    win = True

            if win:
                ticket_total = 0
                for row in ticket:
                    for number in row:
                        if number != 'X':
                            ticket_total += int(number)

                return ticket_total * int(ball)


result = draw_balls()
print(result)
